src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java

```diff
@@ -71,11 +73,26 @@ public SingleClusterMonitor(Map<String, Object> props, String name) throws Excep
   @Override
   public void start() {
     _topicManagementService.start();
-    CompletableFuture<Void> completableFuture = _topicManagementService.topicManagementResult();
-    completableFuture.thenRun(() -> {
+    CompletableFuture<Void> topicPartitionResult = _topicManagementService.topicPartitionResult();
+    try {
+      /* Delay 2 second to reduce the chance that produce and consumer thread has race condition
```

Comment:
I think this approach of using a hardcoded delay to prevent potential race condition is kind of uncommon or not the best practice in my opinion. In general, concurrency problems should NOT just be "reduce the chance" instead it should be solved/prevented completely by using concurrency control primitives (such as CompletableFuture in this case). So I need some clarification to see the purpose of this change.

Isn't the purpose of using the CompletableFuture to establish orders of service starting orders in order to prevent race conditions? I was reading through the code and I don't see why using CompletableFuture is not enough in this case or even if there is some race condition problem, why can't it solved theoretically by using the CompletableFuture.

Reply:
I've discussed this with Sean in the past in this regards. This a similar practice followed by Dong Lin when the ConsumeService was being implemented, in the `private void consume()` method. The TopicManagementService instantiates a new `MultiClusterTopicManagementService` in the constructor, which spins off a new executor thread (this is a separate thread). While the topicPartitionResult waits for the `MultiClusterTopicManagementService` to finish, the new scheduled thread is separate. It's been tested that giving a two second window for the the SingleClusterMonitor to wait is enough time frame to avoid race condition, as is originally done in the `consume()` of `ConsumeService`

```_executor = Executors.newSingleThreadScheduledExecutor(
      r -> new Thread(r, _serviceName + "-multi-cluster-topic-management-service"));```